### ISSUE TITLE
tf.ConditionalAccumulator does not behave as expected

### ISSUE THREAD

**Issue by ylfzr** [tensorflow/tensorflow#24005]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS 10.12.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):  binary
- TensorFlow version (use command below): 1.10 cpu
- Python version: 3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 
- GPU model and memory:

**Describe the current behavior**
Result of the accumulator.take_grad() not predictable if apply more gradients than num_required.
**Describe the expected behavior**

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
```
NUM_REQUIRED = 3

a = np.arange(12).reshape((3, 4)).astype(np.float32)
dataset = tf.data.Dataset.from_tensor_slices((a, a))
dataset = dataset.repeat(1000).batch(1)
iterator = dataset.make_one_shot_iterator()
x, y = iterator.get_next()

predict = tf.layers.dense(x, 4)
loss = tf.nn.l2_loss(predict - y)

optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
trainable_vars = tf.trainable_variables()[1]

grads_vars = optimizer.compute_gradients(loss, var_list=[trainable_vars])
grads = [grad_var[0] for grad_var in grads_vars]
vars = [grad_var[1] for grad_var in grads_vars]
shapes = [grad_var[1].get_shape() for grad_var in grads_vars]

local_step = tf.get_variable('accu_local_step', dtype=tf.int32, initializer=0, trainable=False)
accumulators = [tf.ConditionalAccumulator(tf.float32, shape=shapes[i]) for i in range(len(vars))]
with tf.control_dependencies([tf.assign_add(local_step, 1)]):
    apply_ops = [accumulators[i].apply_grad(grads[i], local_step=local_step) for i in range(len(vars))]
    apply_ops = tf.group(*apply_ops)
grads_mean = [accumulators[i].take_grad(num_required=NUM_REQUIRED) for i in range(len(vars))]
num_acc = [accumulators[i].num_accumulated() for i in range(len(vars))]

# #
sess_local = tf.train.MonitoredSession()
for i in range(100):
    # print(i)
    if (i + 1) % (NUM_REQUIRED + 1) != 0:
        local_step_np, num_acc_np, grad_tmp, _ = sess_local.run([local_step, num_acc, grads, apply_ops])
        # sess_local.run(reset_op)
        print('local_step_np is: {}, num_ncc_np is: {}'.format(local_step_np, num_acc_np))
        print(grad_tmp)
    else:
        # grad_tmp, _ = sess_local.run([grads, apply_ops])
        # print(grad_tmp)
        grad_tmp, num_acc_np, _, grads_mean_np = sess_local.run([grads, num_acc, apply_ops, grads_mean])
        print('num_ncc_np is: {}'.format(num_acc_np))
        print(grad_tmp)
    # num_acc_np = sess_local.run(num_acc)
        print('grads_mean_np:', grads_mean_np)
```

**Other info / logs**
Runing the above code gives result:
local_step_np is: 1, num_ncc_np is: [0]
[array([ 1.9753755,  1.9916878,  0.9389882, -5.003906 ], dtype=float32)]
local_step_np is: 2, num_ncc_np is: [1]
[array([  3.120553 ,   1.0842252,   3.9395618, -11.656693 ], dtype=float32)]
local_step_np is: 3, num_ncc_np is: [2]
[array([  4.26573   ,   0.17676353,   6.940136  , -18.30948   ],
      dtype=float32)]
num_ncc_np is: [0]
[array([ 1.9753755,  1.9916878,  0.9389882, -5.003906 ], dtype=float32)]
grads_mean_np: [array([  3.1205528,   1.0842255,   3.9395618, -11.6566925], dtype=float32)]
local_step_np is: 5, num_ncc_np is: [1]
[array([  3.120553 ,   1.0842252,   3.9395618, -11.656693 ], dtype=float32)]
local_step_np is: 6, num_ncc_np is: [2]
[array([  4.26573   ,   0.17676353,   6.940136  , -18.30948   ],
      dtype=float32)]
local_step_np is: 7, num_ncc_np is: [3]
[array([ 1.9753755,  1.9916878,  0.9389882, -5.003906 ], dtype=float32)]
num_ncc_np is: [4]
[array([  3.120553 ,   1.0842252,   3.9395618, -11.656693 ], dtype=float32)]
grads_mean_np: [array([ 2.8342583,  1.3110911,  3.1894183, -9.993496 ], dtype=float32)]
local_step_np is: 9, num_ncc_np is: [1]
[array([  4.26573   ,   0.17676353,   6.940136  , -18.30948   ],
      dtype=float32)]
local_step_np is: 10, num_ncc_np is: [2]
[array([ 1.9753755,  1.9916878,  0.9389882, -5.003906 ], dtype=float32)]
local_step_np is: 11, num_ncc_np is: [3]
[array([  3.120553 ,   1.0842252,   3.9395618, -11.656693 ], dtype=float32)]
num_ncc_np is: [0]
[array([  4.26573   ,   0.17676353,   6.940136  , -18.30948   ],
      dtype=float32)]
grads_mean_np: [array([  3.1205528,   1.0842254,   3.9395618, -11.656693 ], dtype=float32)]
local_step_np is: 13, num_ncc_np is: [1]
[array([ 1.9753755,  1.9916878,  0.9389882, -5.003906 ], dtype=float32)]
local_step_np is: 14, num_ncc_np is: [2]

grads_mean_np is not the mean of previously applied grad. 

If we just set run 'apply_ops' 'NUM_REQUIRED' times before run 'grads_mean'(take_grads()), ConditionalAccumulator seems to work properly, but when we run different number of 'apply_ops' before(as in the example, we run 'NUM_REQUIRED' + 1 times), the computed grads_mean(averaged grads) becomes mysterious.

**Comment by skye** [comment]
@tanzhenyu are you familiar with ConditionalAccumulator?

**Comment by mohantym** [comment]
Hi @ylfzr!
It seems you are using older versions(1.x versions) of Tensorflow which is not supported any more. Have you checked in[ latest version](https://www.tensorflow.org/api_docs/python/tf/compat/v1/ConditionalAccumulator) yet? Thanks!
